windows/src/desktop/kmshell/main/OnlineUpdateCheck.pas

```diff
@@ -706,4 +975,426 @@ function TOnlineUpdateSharedData.Params: TOnlineUpdateCheckParams;
+        begin
+          SetLength(FParams.Packages,0);
+          for i := Low(ucr.Packages) to High(ucr.Packages) do
+          begin
+            n := kmcom.Packages.IndexOf(ucr.Packages[i].ID);
+            if n >= 0 then
+            begin
+              pkg := kmcom.Packages[n];
+              j := Length(FParams.Packages);
+              SetLength(FParams.Packages, j+1);
+              FParams.Packages[j].NewID := ucr.Packages[i].NewID;
+              FParams.Packages[j].ID := ucr.Packages[i].ID;
+              FParams.Packages[j].Description := ucr.Packages[i].Name;
+              FParams.Packages[j].OldVersion := pkg.Version;
+              FParams.Packages[j].NewVersion := ucr.Packages[i].NewVersion;
+              FParams.Packages[j].DownloadSize := ucr.Packages[i].DownloadSize;
+              FParams.Packages[j].DownloadURL := ucr.Packages[i].DownloadURL;
+              FParams.Packages[j].FileName := ucr.Packages[i].FileName;
+              pkg := nil;
+            end
+            else
+              FErrorMessage := 'Unable to find package '+ucr.Packages[i].ID;
+          end;
+
+          case ucr.Status of
+            ucrsNoUpdate:
+              begin
+                FErrorMessage := ucr.ErrorMessage;
+              end;
+            ucrsUpdateReady:
+              begin
+                FParams.Keyman.OldVersion := ucr.CurrentVersion;
+                FParams.Keyman.NewVersion := ucr.NewVersion;
+                FParams.Keyman.DownloadURL := ucr.InstallURL;
+                FParams.Keyman.DownloadSize := ucr.InstallSize;
+                FParams.Keyman.FileName := ucr.FileName;
+              end;
+          end;
+          { TODO: OU instead of showing a form to update it just has to go and download }
+
+          if (Length(FParams.Packages) > 0) or (FParams.Keyman.DownloadURL <> '') then
+          begin
+          // Set State Flag Update available
+            Result := oucUpdatesAvailable;
+          end
+          else
+          begin  // set State Flag Idle
+            Result := oucNoUpdates
+          end;
+        end
+        else
+        begin
+          FErrorMessage := ucr.ErrorMessage;
+          Result := oucFailure;
+        end;
+      end
+      else
+        raise EOnlineUpdateCheck.Create('Error '+IntToStr(Response.StatusCode));
+    finally
+      Free;
+    end;
+  except
+    on E:EHTTPUploader do
+    begin
+      if (E.ErrorCode = 12007) or (E.ErrorCode = 12029)
+        then FErrorMessage := S_OnlineUpdate_UnableToContact
+        else FErrorMessage := WideFormat(S_OnlineUpdate_UnableToContact_Error, [E.Message]);
+      Result := oucFailure;
+    end;
+    on E:Exception do
+    begin
+      FErrorMessage := E.Message;
+      Result := oucFailure;
+    end;
+  end;
+
+  with TRegistryErrorControlled.Create do  // I2890
+  try
+    if OpenKey(SRegKey_KeymanDesktop_CU, True) then
+      WriteDateTime(SRegValue_LastUpdateCheckTime, Now);
+  finally
+    Free;
+  end;
+end;
```

Comment:
This function is fairly long and has several nested loops and try/except blocks. It might be easier to understand if parts were refactored in separate functions/procedures.